docs/guide/cache/content_aware_cache.md

```diff
@@ -293,3 +350,253 @@ bin/console fos:httpcache:invalidate:tag ez-all
+The second ESI gives a similar response so we'll jump right down to the 3rd ESI
+
+#### 3rd ESI
+
+```bash
+    $ curl -IXGET --resolve www.staging.foobar.com.us-2.platformsh.site:443:1.2.3.4 --header "Surrogate-Capability: abc=ESI/1.0" --header "x-user-hash: daea248406c0043e62997b37292bf93a8c91434e8661484983408897acd93814" 'https://www.staging.foobar.com.us-2.platformsh.site//_fragment?_hash=lnKTnmv6bb1XpaMPWRjV3sNazbn9rDXskhjGae1BDw8%3D&_path=locationId%3D2%26contentId%3D52%26blockId%3D13%26versionNo%3D3%26languageCode%3Deng-GB%26serialized_siteaccess%3D%257B%2522name%2522%253A%2522site%2522%252C%2522matchingType%2522%253A%2522default%2522%252C%2522matcher%2522%253Anull%252C%2522provider%2522%253Anull%257D%26serialized_siteaccess_matcher%3Dnull%26_format%3Dhtml%26_locale%3Den_GB%26_controller%3DEzSystems%255CCustomBundle%255CController%255CFooController%253A%253AcustomAction'
+```
+
+This ESI is handled by a custom `FooController::customAction` and the output of the command is:
+
+Output:
+
+```
+HTTP/1.1 200 OK
+Server: nginx/1.20.0
+Content-Type: text/html; charset=UTF-8
+Transfer-Encoding: chunked
+Connection: keep-alive
+Set-Cookie: eZSESSID21232f297a57a5a743894a0e4a801fc3=asrpqgmh5ll5ssseca3cov8er7; path=/; HttpOnly; SameSite=lax
+Cache-Control: public, s-maxage=86400
+Date: Wed, 01 Sep 2021 07:51:40 GMT
+Vary: Origin
+Vary: X-User-Hash
+Vary: X-Editorial-Mode
+X-Cache-Debug: 1
+Surrogate-Key: ez-all
+```
+
+As you can see, the `Cache-Control` and `Vary` headers looks all good. The request is as mentioned handled by a custom
+controller and the `Surrogate-Key` only contains the default `ez-all` value. That is not a problem as long as the controller
+doesn't return values from any Content in the [[= product_name =]] Repository. If it does, the controller should also add
+the corresponding IDs to such objects in that header.
+
+The big problem here is however the `Set-Cookie`. A ESI fragment should never set a cookie because:
+- Clients will only receive the headers set in the "mother" document (the headers in the "/" response in this case)
+- Only the content of ESIs responses will be returned to the client. *No headers set in the ESI response will ever reach
+  the client*. ESI headers are only seen by the HTTP Cache
+- Symfony reverse proxy doesn't support ESIs at all, and any ESI calls (`render_esi()`) will implicitly be replaced by
+  sub-requests (`render()`). So any `Set-Cookie` *will* be sent to the client when using Symgony reverse proxy.
+- Fastly will flag that resource as "not cachable" since it did set a cookie at least once. Even though that endpoint
+  stops setting cookies, Fastly will still not cache that fragment. Any document referencing that ESI will be a `MISS`.
+  Fastly cache needs to be purged (`Purge-all` request) in order to remove this flag
+- This means that it is a bad idea to always initiate a session when loading the front-page
```

Comment:
```suggestion
The `Set-Cookie`here may cause the problem. A ESI fragment should never set a cookie because:
- Clients will only receive the headers set in the "mother" document (the headers in the "/" response in this case).

- Only the content of ESIs responses will be returned to the client. **No headers set in the ESI response will ever reach the client**. ESI headers are only seen by the HTTP cache.
  
- Symfony reverse proxy does not support ESIs at all, and any ESI calls (`render_esi()`) will implicitly be replaced by
  sub-requests (`render()`). So any `Set-Cookie` **will** be sent to the client when using Symfony reverse proxy.
  
- Fastly will flag it resource as "not cachable" because it set a cookie at least once. Even though that endpoint.
  stops setting cookies, Fastly will still not cache that fragment. Any document referring to that ESI will be a `MISS`.
  Fastly cache needs to be purged (`Purge-all` request) in order to remove this flag.
  
- It means that it is not recommended to always initiate a session when loading the front page.
```

Reply:
FYI :Edited your suggestion : "the `Set-Cookie`here" --> "the `Set-Cookie` here"